lib/identity-document.js

```diff
@@ -0,0 +1,85 @@
+const fs = require('fs');
+const forge = require('node-forge');
+const crypto = require('crypto');
+const assert = require('assert');
+const asn1 = forge.asn1;
+const pkcs7 = forge.pkcs7;
+const oids = forge.oids;
+
+const AWS_PUBLIC_CERTIFICATE =
+`-----BEGIN CERTIFICATE-----
+MIIC7TCCAq0CCQCWukjZ5V4aZzAJBgcqhkjOOAQDMFwxCzAJBgNVBAYTAlVTMRkw
+FwYDVQQIExBXYXNoaW5ndG9uIFN0YXRlMRAwDgYDVQQHEwdTZWF0dGxlMSAwHgYD
+VQQKExdBbWF6b24gV2ViIFNlcnZpY2VzIExMQzAeFw0xMjAxMDUxMjU2MTJaFw0z
+ODAxMDUxMjU2MTJaMFwxCzAJBgNVBAYTAlVTMRkwFwYDVQQIExBXYXNoaW5ndG9u
+IFN0YXRlMRAwDgYDVQQHEwdTZWF0dGxlMSAwHgYDVQQKExdBbWF6b24gV2ViIFNl
+cnZpY2VzIExMQzCCAbcwggEsBgcqhkjOOAQBMIIBHwKBgQCjkvcS2bb1VQ4yt/5e
+ih5OO6kK/n1Lzllr7D8ZwtQP8fOEpp5E2ng+D6Ud1Z1gYipr58Kj3nssSNpI6bX3
+VyIQzK7wLclnd/YozqNNmgIyZecN7EglK9ITHJLP+x8FtUpt3QbyYXJdmVMegN6P
+hviYt5JH/nYl4hh3Pa1HJdskgQIVALVJ3ER11+Ko4tP6nwvHwh6+ERYRAoGBAI1j
+k+tkqMVHuAFcvAGKocTgsjJem6/5qomzJuKDmbJNu9Qxw3rAotXau8Qe+MBcJl/U
+hhy1KHVpCGl9fueQ2s6IL0CaO/buycU1CiYQk40KNHCcHfNiZbdlx1E9rpUp7bnF
+lRa2v1ntMX3caRVDdbtPEWmdxSCYsYFDk4mZrOLBA4GEAAKBgEbmeve5f8LIE/Gf
+MNmP9CM5eovQOGx5ho8WqD+aTebs+k2tn92BBPqeZqpWRa5P/+jrdKml1qx4llHW
+MXrs3IgIb6+hUIB+S8dz8/mmO0bpr76RoZVCXYab2CZedFut7qc3WUH9+EUAH5mw
+vSeDCOUMYQR7R9LINYwouHIziqQYMAkGByqGSM44BAMDLwAwLAIUWXBlk40xTwSw
+7HX32MxXYruse9ACFBNGmdX2ZBrVNGrN9N2f6ROk0k9K
+-----END CERTIFICATE-----`;
+
+// Returns the Identity Document if the signature is valid, null otherwise.
+//
+// References:
+//  - https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instance-identity-documents.html
+//  - http://www.pkiglobe.org/pkcs7.html
+//  - https://tools.ietf.org/html/rfc2315
+module.exports = function(signedData) {
+  let p7;
+
+  try {
+    p7 = pkcs7.messageFromAsn1(asn1.fromDer(forge.util.decode64(signedData)));
+  } catch (err) {
+    console.log(err);
+    return null;
+  }
+
+  const doc = p7.rawCapture.content.value[0].value[0].value;
+  const attrs = p7.rawCapture.authenticatedAttributes;
+
+  const encodedAttrs = asn1.toDer(asn1.create(
+    asn1.Class.UNIVERSAL,
+    asn1.Type.SET,
+    true,
+    attrs
+  ));
+
+  const buf = new Buffer(encodedAttrs.data, 'binary');
+  const digestAlgorithmOid = asn1.derToOid(
+    p7.rawCapture.digestAlgorithm
+  );
+  const signatureAlgorithmOid = asn1.derToOid(
+    p7.rawCapture.signatureAlgorithm[0].value
+  );
+
+  const h = crypto.createHash(forge.oids[digestAlgorithmOid]);
+  const verify = crypto.createVerify(forge.oids[signatureAlgorithmOid]);
+  const messageDigestOid = forge.oids['messageDigest'];
+
+  h.update(doc);
+
+  const messageDigest = attrs.filter(attr => {
+    const attrType = asn1.derToOid(attr.value[0].value);
+    return attrType === messageDigestOid;
+  }).map(attr => attr.value[1].value[0].value)[0];
+
+  verify.update(buf);
+
+  const sig = new Buffer(p7.rawCapture.signature, 'binary');
+  if (h.digest('binary') !== messageDigest
```

Comment:
I was talking to @franziskuskiefer about this and it might be better to not do a second hashing outside of the `verify` instance.

Also, looking at the metadata service, I saw that there's also an RSA2048 version of the signed message.  It's also stored in a PKCS#7 file and has a much stronger hashing algorithm (SHA256 vs SHA1).  We should use that.

Reply:
afaic, if we don't do this hash, we don't validate anything. The `verify` part makes sure `messageDigest`  is valid and hashed by Amazon, while `h.digest` is what actually validates the message. If we skip this part, someone might tamper the Identity Document and we will never know.

Reply:
That's what the verify class does.  In order to verify the signature we need:

1. The document to verify
2. The algorithm to use for verification
3. The public key
4. The signature

These values are provided, respectively, by:

1. Calling `verify.update(buf)`
2. Passing in the correct algorithm to the `crypto.createVerify()` factory
3. Passing to the `verify.verify` function
4. Passing to the `verify.verify` function

The `messageDigest` is actually the hash of the message body, just a different name.  The thing that it's hashing is the document.  Also, it seems to me that the message content being stored in the PKCS7 file is some special thing that amazon does.  This is highlighted by their command to verify the signature requiring you to separately obtain the document and pass it along with the PCKS7 file to the openssl binary.  Let's not rely on this special behaviour.

Let's send a cleartext copy of the document and the PKCS7 file as well and do the verification, as tools expect.

I'm also having trouble getting the unit tests to pass, as well as on travis.  Are you able to run them locally?

Reply:
> The messageDigest is actually the hash of the message body, just a different name. The thing that it's hashing is the document.

Exactly. The signature verification only validates the `authenticatedAttributes` field, which contains the document hash. If we only verify that, we know that the digest in the signature is valid, but not the actual content.

> I'm also having trouble getting the unit tests to pass, as well as on Travis. Are you able to run them locally?

This is due at the time I pushed it, my forge patches weren't merged and I was relying upon an npm released version with patches applied locally. Not sure if we should wait for a release or freeze it master.